### PR TITLE
Implement modification frequency metric

### LINKED ISSUE #189: Metric: Modification Frequency
As an architect, I want to know which parts of the code have been modified how frequently within a certain time frame, in order to identify if any actions are necessary to change the architecture to better support frequent changes.

We should implement this not as a metric plugin, but instead as a core functionality.

### PR DESCRIPTION
**Pull request by KilianKrause**
This is another coderadar-core metric.
Files are declared as critical, if they are edited >= _n_ times within a certain time frame, where _n_ is given by the user.
This PR closes #189

### CODE REVIEW

File: coderadar-graph/src/main/java/io/reflectoring/coderadar/graph/query/adapter/GetCriticalFilesAdapter.java [hunk 1]
```diff
@@ -1,28 +1,65 @@
...
+        contributorQueryRepository.getFilesWithManyContributors(
+            projectId,
+            numberOfContributors,
+            commitHash,
+            includesAndExcludes.getFirst(),
+            includesAndExcludes.getSecond()));
+  }
+
+  @Override
+  public List<FileAndCommitsForTimePeriod> getFrequentlyChangedFiles(
```
Comment (maximAtanasov) [inline coderadar-graph/src/main/java/io/reflectoring/coderadar/graph/query/adapter/GetCriticalFilesAdapter.java]: What happens if the start date is after the commit time?

File: coderadar-core/src/main/java/io/reflectoring/coderadar/query/port/driver/GetCriticalFilesUseCase.java [hunk 2]
```diff
@@ -1,9 +1,13 @@
 package io.reflectoring.coderadar.query.port.driver;
 
-import io.reflectoring.coderadar.contributor.port.driver.GetCriticalFilesCommand;
 import io.reflectoring.coderadar.query.domain.ContributorsForFile;
+import io.reflectoring.coderadar.query.domain.FileAndCommitsForTimePeriod;
 import java.util.List;
 
 public interface GetCriticalFilesUseCase {
-  List<ContributorsForFile> getCriticalFiles(Long projectId, GetCriticalFilesCommand command);
+  List<ContributorsForFile> getFilesWithManyContributors(
```
Comment (maximAtanasov) [inline coderadar-core/src/main/java/io/reflectoring/coderadar/query/port/driver/GetCriticalFilesUseCase.java]: Please add javadoc to this interface.

File: coderadar-core/src/main/java/io/reflectoring/coderadar/query/port/driven/GetCriticalFilesPort.java [hunk 3]
```diff
@@ -2,9 +2,18 @@
...
 import io.reflectoring.coderadar.projectadministration.domain.FilePattern;
 import io.reflectoring.coderadar.query.domain.ContributorsForFile;
+import io.reflectoring.coderadar.query.domain.FileAndCommitsForTimePeriod;
+import java.util.Date;
 import java.util.List;
 
 public interface GetCriticalFilesPort {
-  List<ContributorsForFile> getCriticalFiles(
-      Long projectId, int numberOfContributors, String commitHash, List<FilePattern> filePatterns);
+  List<ContributorsForFile> getFilesWithManyContributors(
```
Comment (maximAtanasov) [inline coderadar-core/src/main/java/io/reflectoring/coderadar/query/port/driven/GetCriticalFilesPort.java]: Please add javadoc to this interface.

File: coderadar-core/src/main/java/io/reflectoring/coderadar/projectadministration/domain/Commit.java [hunk 4]
```diff
@@ -8,15 +9,15 @@
 @Data
 public class Commit {
 
-  private long id;
+  @JsonIgnore private long id;
```
Comment (maximAtanasov) [inline coderadar-core/src/main/java/io/reflectoring/coderadar/projectadministration/domain/Commit.java]: I know we discussed this part. I just wanted to ask if you've tried doing this with mapping? Just curious if it's noticeably slower. If it performs exactly the same, then maybe we could consider mapping.

**Comment by maximAtanasov** [comment]
@KilianKrause 

> Files are declared as critical, if they are edited n times within a certain time frame, where n is given by the user.

You probably mean `more than n times`?

File: coderadar-ui/package.json [hunk 5]
```diff
@@ -3,7 +3,7 @@
   "version": "0.2.3",
   "scripts": {
     "ng": "node --max_old_space_size=5048 ./node_modules/@angular/cli/bin/ng",
-    "start": "node --max_old_space_size=5048 ./node_modules/@angular/cli/bin/ng serve --prod",
+    "start": "node --max_old_space_size=5048 ./node_modules/@angular/cli/bin/ng serve",
```
[the remark this answers is not in the document]
Comment (KilianKrause) [reply coderadar-ui/package.json]: Done

File: coderadar-graph/src/main/java/io/reflectoring/coderadar/graph/query/repository/ContributorQueryRepository.java [hunk 6]
```diff
@@ -22,7 +22,7 @@
           + "UNWIND emails as email "
           + "MATCH (co:ContributorEntity) WHERE toLower(email) IN co.emails WITH path, collect(DISTINCT co.displayName) as contributors "
           + "WHERE size(contributors) = {1} RETURN path, contributors")
-  List<ContributorsForFileQueryResult> getCriticalFiles(
+  List<ContributorsForFileQueryResult> getFilesWithManyContributors(
```
[the remark this answers is not in the document]
Comment (KilianKrause) [reply coderadar-graph/src/main/java/io/reflectoring/coderadar/graph/query/repository/ContributorQueryRepository.java]: Done

File: coderadar-graph/src/main/java/io/reflectoring/coderadar/graph/query/adapter/GetCriticalFilesAdapter.java [hunk 7]
```diff
@@ -1,28 +1,65 @@
...
+            projectId,
+            commitHash,
+            startDate.getTime(),
+            frequency,
+            includesAndExcludes.getFirst(),
+            includesAndExcludes.getSecond()));
+  }
 
-    // Map Ant-Patterns to RegEx
+  private Pair<List<String>, List<String>> mapPatternsToRegex(List<FilePattern> filePatterns) {
```
[the remark this answers is not in the document]
Comment (KilianKrause) [reply coderadar-graph/src/main/java/io/reflectoring/coderadar/graph/query/adapter/GetCriticalFilesAdapter.java]: Done

File: coderadar-core/src/main/java/io/reflectoring/coderadar/query/service/GetCriticalFilesService.java [hunk 8]
```diff
@@ -28,16 +30,33 @@ public GetCriticalFilesService(
...
+    List<FilePattern> filePatterns = checkProjectExistsAndGetFilePatterns(projectId);
+    return getCriticalFilesPort.getFrequentlyChangedFiles(
+        projectId,
+        command.getCommitHash(),
+        command.getStartDate(),
+        command.getFrequency(),
+        filePatterns);
+  }
+
+  private List<FilePattern> checkProjectExistsAndGetFilePatterns(long projectId) {
```
[the remark this answers is not in the document]
Comment (KilianKrause) [reply coderadar-core/src/main/java/io/reflectoring/coderadar/query/service/GetCriticalFilesService.java]: Done

File: coderadar-core/src/main/java/io/reflectoring/coderadar/query/port/driver/GetCriticalFilesUseCase.java (hunk 2 above)
Comment (KilianKrause) [reply coderadar-core/src/main/java/io/reflectoring/coderadar/query/port/driver/GetCriticalFilesUseCase.java]: Done

File: coderadar-core/src/main/java/io/reflectoring/coderadar/query/port/driven/GetCriticalFilesPort.java (hunk 3 above)
Comment (KilianKrause) [reply coderadar-core/src/main/java/io/reflectoring/coderadar/query/port/driven/GetCriticalFilesPort.java]: Done

File: coderadar-graph/src/main/java/io/reflectoring/coderadar/graph/query/adapter/GetCriticalFilesAdapter.java (hunk 1 above)
Comment (KilianKrause) [reply coderadar-graph/src/main/java/io/reflectoring/coderadar/graph/query/adapter/GetCriticalFilesAdapter.java]: Nothing is returned. The date of the commits need to be at a later time than start date. We could declare this as a misconfiguration and tell the user that the input is invalid.

File: coderadar-rest/src/main/java/io/reflectoring/coderadar/rest/query/GetCriticalFilesController.java [hunk 9]
```diff
@@ -23,9 +25,18 @@ public GetCriticalFilesController(GetCriticalFilesUseCase getCriticalFilesUseCas
   @GetMapping(
```
[the remark this answers is not in the document]
Comment (KilianKrause) [reply coderadar-rest/src/main/java/io/reflectoring/coderadar/rest/query/GetCriticalFilesController.java]: Done

File: coderadar-rest/src/main/java/io/reflectoring/coderadar/rest/query/GetCriticalFilesController.java [hunk 10]
```diff
@@ -23,12 +26,52 @@ public GetCriticalFilesController(GetCriticalFilesUseCase getCriticalFilesUseCas
...
+    for (FileAndCommitsForTimePeriod f : files) {
+      FileAndCommitsForTimePeriodResponse resultItem = new FileAndCommitsForTimePeriodResponse();
+      resultItem.setPath(f.getPath());
+      resultItem.setCommits(mapCommits(f.getCommits()));
+      result.add(resultItem);
+    }
+    return result;
+  }
+
+  private List<GetCommitResponse> mapCommits(List<Commit> commits) {
```
Comment (maximAtanasov) [inline coderadar-rest/src/main/java/io/reflectoring/coderadar/rest/query/GetCriticalFilesController.java]: You now have the same method in different classes. Maybe extract it to a `GetCommitResponseMapper` or something similar.

Comment (KilianKrause) [reply coderadar-rest/src/main/java/io/reflectoring/coderadar/rest/query/GetCriticalFilesController.java]: Done

**Pull request merged**